Add `calc_displacements_ellipsoid` function

kavanase opened the pull request.

kavanase commented:
@coderabbitai review

kavanase commented:
@teruya7 this looks very nice, thank you very much for adding to the code! 🙌 

Some small things on the checklist to get this merged:
- Would you mind adding some quick test cases, similar to those added for the original `calc_site_displacements` functions in `test_displacements.py`?
    - Likewise, adding a short example (probably using one of these test cases) to the [advanced analysis tutorial](https://doped.readthedocs.io/en/latest/advanced_analysis_tutorial.html#defect-induced-site-displacements-strain) would be nice to illustrate this feature!
- (Lower priority): It would be nice if, similar to `plot_site_displacements`, using `plotly` instead of `matplotlib` was an option rather than a requirement. Would you be able to try implementing both backends?

kavanase commented:
Also following the CodeRabbit suggestions above; could you change the `dict()` calls to literal dicts (`{...}`) as suggested? This is the preferred formatting if you use the `doped` `pre-commit`

teruya7 commented:
@kavanase 
Thank you for your kind comments.
I will modify the points you have mentioned.

kavanase commented:
@teruya7 that was quick! 🙌  
I have given you write access to `doped` now, so would be easiest to push directly to this PR for making this as efficient as possible.
Can you make sure to pull this branch as is first before making any other updates, to avoid merge conflicts? (There were a good few with the last branch so this will avoid some headaches).

kavanase commented on tests/test_displacements.py:
```diff
@@ -126,6 +126,22 @@ def test_plot_site_displacements_error(self):
         with pytest.raises(ValueError):
             defect_entry.plot_site_displacements(vector_to_project_on=[0, 0, 1], relative_to_defect=True)
 
+    def test_calc_displacements_ellipsoid(self):
```
Nice! 
Could you add some quick image comparison tests (for the plotting functions, using `matplotlib`) here too, like the tests at the bottom of this module?
For this you just need to add `@custom_mpl_image_compare(filename="...")` above the test function as for the others, where `filename` is the name of the reference file to compare to in `tests/data/remote_baseline_plots`, and your test function returns the `matplotlib` `Figure` object. Would be good to do some quick tests for the different possible combinations of options.

Then maybe a quick test each like you've done here for the outputs of the function with a substitution and interstitial defect please? (Best to test each type of defect when possible). There's an example of this in `test_calc_site_displacements`

kavanase commented:
Added some other fixes and cleanup there. One issue, for the cell axes in the `plotly` outputs, there's a small issue where it seems to be getting them a bit wrong:
![image](https://github.com/user-attachments/assets/b1d270df-70a6-42bf-b76e-e59768fc1b57)
(From the tutorial notebook plot)

I think this is probably due to my refactoring of that part of the code to try make it a bit more succinct. Would you be able to find the issue with this and fix it? 🙏

kavanase commented on doped/utils/displacements.py:
```diff
@@ -553,3 +489,726 @@ def _plotly_plot_total_disp(
+                site.frac_coords, sc_defect_frac_coords, bulk_sc
+            )[1]
+        )
+        disp_dict["Z sites in cartesian coordinate (defect)"].append(
+            _shift_defect_site_to_center_of_the_supercell(
+                site.frac_coords, sc_defect_frac_coords, bulk_sc
+            )[2]
+        )
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "invalid value encountered in scalar divide")
+
+    # sort each list in disp dict by index of species in bulk element list, then by distance to defect:
+    element_list = [
+        el.symbol for el in defect_entry.defect.structure.composition.elements
+    ]  # host elements
+    element_list += sorted(
+        [  # extrinsic elements, sorted alphabetically for deterministic ordering in output:
+            el.symbol
+            for el in defect_entry.defect.defect_structure.composition.elements
+            if el.symbol not in element_list
+        ]
+    )
+    # Combine the lists into a list of tuples, then sort, then unpack:
+    combined = list(zip(*disp_dict.values()))
+    combined.sort(
+        key=lambda x: (element_list.index(x[1]), x[4], x[0])
+    )  # Sort by species, then distance, then index
+
+    (
+        disp_dict["Index (defect)"],
+        disp_dict["Species"],
+        disp_dict["Species_with_index"],
+        disp_dict["Displacement"],
+        disp_dict["Displacement Norm"],
+        disp_dict["Distance to defect"],
+        disp_dict["X sites in cartesian coordinate (defect)"],
+        disp_dict["Y sites in cartesian coordinate (defect)"],
+        disp_dict["Z sites in cartesian coordinate (defect)"],
+    ) = zip(*combined)
```
A good chunk of this code is duplicated from the `calc_site_displacements` function. It would be best for code efficiency and maintenance if this is pulled out into a helper function that then is used in both functions (with flags for `relative_to_defect`, `vector_to_project_on` and `ellipsoid` to determine which `disp_dict` data is needed?).
Could you implement this please?

kavanase reviewed (commented):
Added two code requests there!

Once done, I think this is all ready to be merged 😃 
Thanks again @teruya7!

kavanase commented:
I also updated the functions to return the `Figure` objects if generated btw, which is useful so users can further customise if they want, and for testing.

teruya7 commented:
Thank you for checking!
I will modify codes.

kavanase commented:
Hi @teruya7!
Just checking in, we're hoping to release a new major version of `doped` in the coming weeks, and would be nice to have this functionality included. Would you have time to work on the remaining todos from the code review above over the next while? 🙏 Thanks again for these additions!

teruya7 commented:
@kavanase 
I have been busy lately and have not been able to update the code. I will update by tomorrow.

teruya7 commented:
@kavanase 
I'm sorry for the delay.
I have corrected the things you pointed out.

kavanase commented:
@teruya7, no problem at all!
Thanks for your work on this. Can you push these updates? (There are no new commits showing on this PR)

kavanase commented:
@teruya7 there were some merge conflicts with the `develop` branch, so I have fixed these now and pushed to this PR. I think you will need to pull the current PR with `git pull --rebase`, and then `git push` your updates. Thanks!

teruya7 commented:
@kavanase 
I pulled the current PR and then pushed my update. Could you check whether the merge conflict is solved or not?

kavanase commented:
Thanks for your work on this @teruya7!
I've gone through and cleaned up the code a bit and added some more tests. Merging to `develop` now!
This will be included in the next `doped` release, which should be in the coming weeks 🤞
